RoboJackets/robocup-software
# Rrt off lines

## Description

EDIT: fix one issue, create more issues...do not review this yet

During the last scrim we noticed issues where our RRT-based planner took unnecessarily wide lines around obstacles. This was especially obvious in our wall, when sometimes robots that only needed to move in a couple straight line segments would take very long paths, leaving holes in our wall to be scored on.

This PR addresses that by making a piecewise trajectory around the first obstacle before resorting to full RRT. Since this trajectory is found deterministically and tries to cut as narrow a path as possible, it should result in better motion planning behavior in crowded situations than RRT. [See this comment for more details.](https://github.com/RoboJackets/robocup-software/pull/1890/files#diff-dd3f19ffa77540702382cf850fceac2bb94109f100445678138af56441a8985fR12)

## Steps to test
### Test Case 1
1. Run normal plays and see what happens.

Expected result: soccer should not crash, robots move to points on cleaner lines.

## Test Procedure

All timings are based on the time given by our UI (see screen recordings). Timer starts when blue UI trajectories are drawn and stops when yellow UI trails disappear (indicating no more movement). If a hole in the wall is left, stop the timer but note that.

1. Place ball on intersection of right sideline and midfield line, wait for robots to form wall.
3. Place ball on left sideline at x=3.0, y=2, time it
4. Repeat 1
5. Place opp robot at x=0, y=1.2
6. Repeat 2, time it

## Test Results

In general, the robots move more cleanly around the opponents, but are more likely to run into each other. A 10-man wall is used here to stress-test: with fewer robots both planners look better. The crashing issue (for wallers at least) can be resolved with a dedicated wall planner, where each robot is given the trajectory along the walling arc instead of a straight line to their destination.

[`ros2`](https://drive.google.com/file/d/13eyAMNUk51dmign3BCn-prh2NY1BfRfq/view?usp=sharing)
2) 7.57 s
4) 7.53 s
5) 9.40 s
**avg: 8.17 s**
every trial lost 1 robot from the wall

[`rrt-off-lines`](https://drive.google.com/file/d/15UBhRpwILHyaSh3us_86r78M4Z0FSmmn/view?usp=sharing)
2) 5.73 s
4) 3.66 s
5) 4.48 s
**avg: 4.62 s**
no loss of wallers in any trial
